[AIRFLOW-5920] Neo4j operator and hook

tfindlay-au opened the pull request:
Make sure you have checked _all_ steps below.

### Jira
AIRFLOW-5920
See:
https://issues.apache.org/jira/browse/AIRFLOW-5920

### Description
These changes add the ability for a pipeline to execute a cypher query as part of a DAG pipeline.

The query is executed against an Airflow managed connection (host, port, username, password) using the Neo4j python library.

### Tests

Unit tests have been submitted

### Commits



### Documentation

Documentation supplied.

tfindlay-tw committed:
AIRFLOW-5920: First draft, added hook + operator

tfindlay-au committed:
AIRFLOW-5920: Tidy up, added option to fail when no results

tfindlay-tw committed:
AIRFLOW-5920: Adding first draft of documentation

tfindlay-tw committed:
AIRFLOW-5920: Added unit tests
AIRFLOW-5920: Refactored neo4j_hook to be stateless, can adapt to a pooled connection in the future.

tfindlay-tw committed:
Merge branch 'master' of https://github.com/apache/airflow

 Conflicts:
	docs/operators-and-hooks-ref.rst

tfindlay-tw committed:
AIRFLOW-5920: Updated documentation

turbaszek commented on airflow/contrib/hooks/neo4j_hook.py:
```diff
@@ -0,0 +1,93 @@
+        """
+        a = Neo4JHook.get_config(self._n4j_conn_id)
+        b = Neo4JHook.get_driver(a)
+        c = Neo4JHook.get_session(b)
```
Please use more meaningful variables' names.

tfindlay-tw replied:
Yes, apologies, I did name them better initially, but during testing I did this, they should be fixed up now.

turbaszek commented on airflow/contrib/operators/neo4j_operator.py:
```diff
@@ -0,0 +1,92 @@
+        if self._cypher_query is not None:
+            result = hook.run_query(cypher_query=self._cypher_query)
+        else:
+            raise AirflowException("cypher_query is missing.")
```
This try / except is unnecessary because `cypher_query` is required in constructor. It's also good idea to skip `_`in attributes. I would also suggest adding `template_fields` to the operator.

tfindlay-tw replied:
Sure, at one point, I considered making it optional. Queries could be read in from files or other sources, but then simplified it to just accept a string. Have removed this to reflect how it is designed to operate right now. Complexity can be added later if needed.

tfindlay-au replied:
When you say, good idea to skip the `_` (underscore) in attributes, are you refering to the class variables ?
I understood this was the convention for private variables. Happy to change it if you think it should be changed.

turbaszek commented on airflow/contrib/operators/neo4j_operator.py:
```diff
@@ -0,0 +1,92 @@
+            output_writer = csv.DictWriter(output_file, fieldnames=result.keys())
+            output_writer.writeheader()
+
+            for total_row_count, row in enumerate(result, start=1):
```
Is there any reason for the enumerate?

tfindlay-tw replied:
Only to provide a counter to how many records were processed. It would be good if there was a more efficient way of doing it, but I dont believe the BoltStatementResult provides a count for read queries.
Happy to remove if its of a concern, I just felt like feedback can be useful and this was a way of capturing something.

turbaszek commented on docs/howto/operator/neo4j.rst:
```diff
@@ -0,0 +1,72 @@
+    :language: python
+    :dedent: 4
+    :start-after: [START howto_operator_dingding]
+    :end-before: [END howto_operator_dingding]
```
Those start/end will need to be adjusted to build docs.

tfindlay-tw replied:
Opps, this should be fixed now.

turbaszek reviewed (changes requested):
I left a few comments. As it's still WIP I would suggest adding type-hints, they help a lot ;)

tfindlay-tw committed:
AIRFLOW-5920: Resolved pylint issues
AIRFLOW-5920: Corrected RST link
AIRFLOW-5920: Added type hints as suggested.

tfindlay-tw committed:
AIRFLOW-5920: Resolved documentation build issues
AIRFLOW-5920: Added dependency to setup.py

tfindlay-tw committed:
AIRFLOW-5920: Resolved all flake8 + pylint + mypy issues.

tfindlay-tw committed:
AIRFLOW-5920: Fixed isort issue
AIRFLOW-5920: Fixed pydoc issue in neo4j_hook

turbaszek commented on airflow/contrib/hooks/neo4j_hook.py:
```diff
@@ -0,0 +1,106 @@
+                config['credentials'] = connection_object.login, connection_object.password
+                config['host'] = "bolt://{0}:{1}".format(connection_object.host, connection_object.port)
+        else:
+            raise AirflowException("No Neo4J connection: {}".format(n4j_conn_id))
```
I would do this like:
```python
if n4j_conn_id is None:
    raise AirflowException("No Neo4J connection: {}".format(n4j_conn_id))

connection_object = Neo4JHook.get_connection(n4j_conn_id)
if connection_object.login and connection_object.host:
    ...
```
It's simpler to read imho. Additionally, will the exception always be formatted with `None`?

tfindlay-tw replied:
I have resolved this by making it require a string, `None` is not accepted by the function. As a result, no exception is to be thrown.

turbaszek commented on airflow/contrib/hooks/neo4j_hook.py:
```diff
@@ -0,0 +1,106 @@
+        return config
+
+    @staticmethod
+    def get_driver(config: dict):
```
```suggestion
    def get_driver(config: dict) -> neo4j.Driver:
```

turbaszek commented on airflow/contrib/hooks/neo4j_hook.py:
```diff
@@ -0,0 +1,106 @@
+        )
+
+    @staticmethod
+    def get_session(driver: Driver):
```
```suggestion
    def get_session(driver: Driver) -> neo4j.Session:
```

turbaszek commented on airflow/contrib/hooks/neo4j_hook.py:
```diff
@@ -0,0 +1,106 @@
+        # Check if we already have a session we can re-use before creating a new one
+        return driver.session()
+
+    def run_query(self, cypher_query: str, parameters=None):
```
```suggestion
    def run_query(self, cypher_query: str, parameters=None) -> neo4j.BoltStatementResult:
```

tfindlay-tw committed:
Merge branch 'master' of https://github.com/apache/airflow

tfindlay-tw committed:
AIRFLOW-5920: Fixed isort issue (with silly spacing)
AIRFLOW-5920: Removed Optional[str] for n4j_conn_id in get_config()
AIRFLOW-5920: Added type hints for method return types as suggested

tfindlay-tw committed:
AIRFLOW-5920: Fixed unused import for flake8

tfindlay-tw committed:
Merge branch 'master' of https://github.com/apache/airflow

tfindlay-tw committed:
AIRFLOW-5920: Add neo4j to dependency tree to fix the build

OmerJog commented on airflow/contrib/example_dags/example_neo4j_operator.py:
```diff
@@ -0,0 +1,60 @@
+)
+# [END howto_operator_neo4j]
+
+t2.set_upstream(t1)
```
Please use bitshift 
https://airflow.apache.org/concepts.html#bitshift-composition

OmerJog commented on airflow/contrib/operators/neo4j_operator.py:
```diff
@@ -0,0 +1,107 @@
+
+class Neo4JOperator(BaseOperator):
+    """
+    This operator takes a number of configuration parameters
```
Need to be furthered explained.
takes number of configuration parameters and.. do what?

tfindlay-au replied:
Fair point, it feels like there is some overlap with the package description. I've tried to make this a bit clearer.

OmerJog commented on airflow/contrib/operators/neo4j_operator.py:
```diff
@@ -0,0 +1,107 @@
+    :type output_filename: str
+    :param n4j_conn_id: reference to a pre-defined Neo4J Connection
+    :type n4j_conn_id: str
+    :param fail_on_no_results: True/False flag to indicate if it should fail the task if no results
```
mmm maybe better name soft_fail? WDYT?

tfindlay-au replied:
Sure, sounds more succinct, I think with a description it should be clear what effect it has.

OmerJog commented on airflow/contrib/operators/neo4j_operator.py:
```diff
@@ -0,0 +1,107 @@
+    _n4j_conn_id = None
+    _fail_on_no_results = None
+
+    template_fields = ['cypher_query', 'output_filename', 'n4j_conn_id']
```
Can we also have `template_ext` for the query ? This will be used to have the query saved on file rather than coded in the operator. Might not be that easy as there is no designated extension https://stackoverflow.com/questions/15161221/neo4j-script-file-format-is-there-any WDYT?

tfindlay-au replied:
Definately worth looking at getting queries from a file. I originally considered this, but simplified to only coded in the operator. But If you think there is a good use, I'll look at adding this in.

tfindlay-au replied:
Done

OmerJog commented on airflow/contrib/operators/neo4j_operator.py:
```diff
@@ -0,0 +1,107 @@
+        # Provide some feedback to what was done...
+        self.log.info("Saved %s with %s rows", self._output_filename, row_count)
+
+    def _make_csv(self, result: BoltStatementResult):
```
Shouldn't this method be in the hook?

tfindlay-au replied:
Excellent point. I think of the hook as providing the descrete mechanisim to connect to the database. It is a wrapper for the library that does this.
There are a few other operators that handle file writing operations like vertica_to_mysql and sql_to_gcs and so on. Happy to move it, but the hook didnt feel like the right place for it either.

OmerJog replied:
Think of case where you would want to export neo4j data (for backup or other)
now you can export it to local file, GCS, S3, FTP and the list is long.
If this function remains on the Operator each one will need to duplicate the logic.
On the other hand if the function is on the hook all operators that use the hook will be able to use it.

This is why most of the db hook has this method in them (and other like get_pandas_df )

By the way most hooks call this method `to_csv` better to stay consistent.

tfindlay-au replied:
Thankyou @OmerJog 
I have moved this from the operator to the hook as you suggested. I could not find any other examples of this, but happy to take your advice on board.

tfindlay-tw committed:
Merge branch 'master' of https://github.com/apache/airflow

tfindlay-tw committed:
AIRFLOW-5920: Updated example DAG to use bitshift operators
AIRFLOW-5920: Updated operator to allow supply of filename for input query
AIRFLOW-5920: Updated operator parameter name from `fail_on_no_results` to `soft_fail`

turbaszek commented on airflow/contrib/operators/neo4j_operator.py:
```diff
@@ -0,0 +1,115 @@
+    _n4j_conn_id = None
+    _soft_fail = None
+
+    template_fields = ['cypher_query', 'output_filename', 'n4j_conn_id', 'soft_fail']
```
This won't work. Template fields should be names of instance attributes that should be rendered. So in this case:
```python
template_fields = ['_cypher_query', '_output_filename', _'n4j_conn_id', '_soft_fail']
```
That's why I suggested removing the underscore.

tfindlay-au replied:
Thanks, I understand now. I've removed the underscore prefix from the class variables now.
Running the tests locally, if it all passes I'll push the change up shortly.
Thanks for clarifying.

tfindlay-tw committed:
AIRFLOW-5920: Fixed parameter name fail_on_no_results -> soft_fail
AIRFLOW-5920: Renamed class variables in the operator

OmerJog commented on airflow/contrib/example_dags/example_neo4j_operator.py:
```diff
@@ -0,0 +1,60 @@
+    "retry_delay": timedelta(minutes=5),
+}
+
+dag = DAG("neo4j_example", default_args=default_args, schedule_interval="@once")
```
Can you change to context manager?
https://airflow.apache.org/concepts.html#context-manager

tfindlay-au replied:
Sure, done.

OmerJog commented on airflow/contrib/operators/neo4j_operator.py:
```diff
@@ -0,0 +1,115 @@
+    :type output_filename: str
+    :param n4j_conn_id: reference to a pre-defined Neo4J Connection
+    :type n4j_conn_id: str
+    :param soft_fail: True/False flag to indicate if it should fail the task if no results
```
Not sure it's clear.. too many ifs. and the True/False flag part is obvious due to the type bool 
Maybe something like:
set True to fail when query return no result.

tfindlay-au replied:
Thanks, that is better. Updated.

OmerJog commented on airflow/contrib/hooks/neo4j_hook.py:
```diff
@@ -0,0 +1,102 @@
+from airflow.hooks.base_hook import BaseHook
+
+
+class Neo4JHook(BaseHook):
```
Would it make sense to inherit from DbApiHook ?

tfindlay-au replied:
I considered this, however cypher queries are not really sql.
Also I was not using a Neo4j connection type and wanted to avoid the complexity of using a ConnectorProtocol or getting into SQL Alchemy.
Whilst Neo4j can be treated like a database, it does not feel like a natural fit for DbApiHook.

tfindlay-tw committed:
AIRFLOW-5920: Changed example DAG to use bitshift and add an example for loading query from file
AIRFLOW-5920: Removed the call to super().__init__() in the neo4j hook
AIRFLOW-5920: Removed the underscore prefix in the hook to make the parameter compatible with template_fields

tfindlay-tw committed:
Merge branch 'master' of https://github.com/apache/airflow

tfindlay-tw committed:
AIRFLOW-5920: Fix RST formatting

feluelle commented on airflow/contrib/hooks/neo4j_hook.py:
```diff
@@ -0,0 +1,101 @@
+    """This class enables the neo4j operator to execute queries against a configured neo4j server.
+    It requires the configuration name as set in Airflow -> Connections ->
+    :param n4j_conn_id:
+    :type str:
```
```suggestion
    :type n4j_conn_id: str
```

feluelle commented on airflow/contrib/hooks/neo4j_hook.py:
```diff
@@ -0,0 +1,101 @@
+class Neo4JHook(BaseHook):
+    """This class enables the neo4j operator to execute queries against a configured neo4j server.
+    It requires the configuration name as set in Airflow -> Connections ->
+    :param n4j_conn_id:
```
Please add a sentence to it.

tfindlay-au replied:
Done.

feluelle commented on airflow/contrib/hooks/neo4j_hook.py:
```diff
@@ -0,0 +1,101 @@
+
+
+class Neo4JHook(BaseHook):
+    """This class enables the neo4j operator to execute queries against a configured neo4j server.
```
In case of multi-line comments please use:
```
"""
foo
"""
```

tfindlay-au replied:
Thanks, have updated to reflect this.

feluelle commented on airflow/contrib/hooks/neo4j_hook.py:
```diff
@@ -0,0 +1,101 @@
+    :param n4j_conn_id:
+    :type str:
+    """
+    n4j_conn_id: str
```
why did you define this here?

tfindlay-au replied:
I defined this class variable as it was set when the class was instantiated via __init__(). It needs to be held in the class object so it can be consumed later.

If there is a better way to store / retrieve this value, I am happy to change this.

tfindlay-au replied:
@feluelle Waiting for your advice on this please. Where would you have defined this ?

feluelle commented on airflow/contrib/hooks/neo4j_hook.py:
```diff
@@ -0,0 +1,101 @@
+
+class Neo4JHook(BaseHook):
+    """This class enables the neo4j operator to execute queries against a configured neo4j server.
+    It requires the configuration name as set in Airflow -> Connections ->
```
An empty line is required after the doc string title.

tfindlay-au replied:
Thanks, should be fixed now.

feluelle commented on airflow/contrib/hooks/neo4j_hook.py:
```diff
@@ -0,0 +1,101 @@
+        :return Session Neo4J session which may contain many transactions
+        :rtype neo4j.Session
+        """
+        # Check if we already have a session we can re-use before creating a new one
```
Where do you check this?

tfindlay-au replied:
Thanks - It originally worked this way, but not any more. I should have removed the comment when refactoring.

feluelle commented on airflow/contrib/hooks/neo4j_hook.py:
```diff
@@ -0,0 +1,101 @@
+        :return Driver connection
+        :rtype neo4j.Driver
+        """
+        # Check if we already have a driver we can re-use before creating a new one
```
Where do you check this?

tfindlay-au replied:
Thanks - It originally worked this way, but not any more. I should have removed the comment when refactoring.

feluelle commented on airflow/contrib/operators/neo4j_operator.py:
```diff
@@ -0,0 +1,115 @@
+
+        row_count = self._make_csv(result)
+
+        # Provide some feedback to what was done...
```
```suggestion
```

I don't think this comment is necessary.

feluelle commented on tests/contrib/hooks/test_neo4j_hook.py:
```diff
@@ -0,0 +1,91 @@
+        assert mock_get_config.called
+        assert mock_get_driver.called
+        assert mock_get_session.called
+        # assert result.assert_called_once_with(lambda tx, inputs: tx.run("QUERY", inputs), None)
```
It would be better to use `mock_obj.assert_called_once_with` instead of `assert mock_obj.called`. You only want it to be called once, don't you?

feluelle reviewed (changes requested):
Please check all you docs:
* An empty line is needed in-between title and params
* consistent type definitions in function header and doc string would be great

You may also want to consider the new upcoming module structure via the `providers` package. See [AIP-21](https://cwiki.apache.org/confluence/display/AIRFLOW/AIP-21%3A+Changes+in+import+paths) for details.

tfindlay-au committed:
Update airflow/contrib/hooks/neo4j_hook.py

Co-Authored-By: Felix Uellendall <feluelle@users.noreply.github.com>

tfindlay-au committed:
Update airflow/contrib/operators/neo4j_operator.py

Co-Authored-By: Felix Uellendall <feluelle@users.noreply.github.com>

tfindlay-tw committed:
AIRFLOW-5920: Improved pydoc comments per recommendations
AIRFLOW-5920: Improved assertions in unit tests

tfindlay-tw committed:
Merge branch 'master' of https://github.com/apache/airflow

 Conflicts:
	setup.py

tfindlay-tw committed:
AIRFLOW-5920: Improved pydoc comments per recommendations
AIRFLOW-5920: Improved assertions in unit tests
AIRFLOW-5920: Rebased from master, re-add neo4j dependency

tfindlay-tw committed:
Merge branch 'master' of https://github.com/apache/airflow

tfindlay-tw committed:
AIRFLOW-5920: Fixed the order in setup.py

tfindlay-tw committed:
Merge branch 'master' of https://github.com/apache/airflow

tfindlay-tw committed:
AIRFLOW-5920: moved & renamed _make_csv() -> to_csv()

tfindlay-tw committed:
Merge branch 'master' of https://github.com/apache/airflow

tfindlay-tw committed:
Merge branch 'master' of https://github.com/apache/airflow

tfindlay-tw committed:
AIRFLOW-5920: Added a line to trigger a build

tfindlay-tw committed:
AIRFLOW-5920: Moved everything to comply with AIP-21

tfindlay-tw committed:
Merge branch 'master' of https://github.com/apache/airflow

tfindlay-tw committed:
AIRFLOW-5920: Moved everything to comply with AIP-21

tfindlay-tw committed:
AIRFLOW-5920: Resolve issue of missing toctree for neo4j

tfindlay-tw committed:
Merge branch 'master' of https://github.com/apache/airflow

tfindlay-tw committed:
AIRFLOW-5920: Fixed the mocks to reflect the refactoring from contrib to providers

ashb commented on airflow/providers/neo4j/hooks/neo4j_hook.py:
```diff
@@ -0,0 +1,138 @@
+# -*- coding: utf-8 -*-
```
@potiuk I can't find our latest "rules" for new naming. Should this be ` airflow/providers/neo4j/hooks/neo4j.py`

turbaszek replied:
I think it's the way it should be. Other option I see is `airflow/providers/neo4j/hooks.py` but this can be confusing.

ashb replied:
```
from airflow.providers.neo4j.hooks.neo4j import Neo4jHook
```

or how about this?

```
from airflow.providers.neo4j.hook import Neo4jHook
```

(`airflow/providers/neo4j/hook.py`) If it's unlikely we'll ever have more than single hook?

turbaszek replied:
I am ok with `airflow/providers/neo4j/hook.py`, @potiuk?

ashb commented on airflow/providers/neo4j/example_dags/__init__.py:
```diff
@@ -0,0 +1,16 @@
+# Licensed to the Apache Software Foundation (ASF) under one
```
Do we need this file? `example_dags` is never "imported" so we don't need to make this a module.

Is this what we do for other providers?

ashb commented on airflow/providers/neo4j/hooks/neo4j_hook.py:
```diff
@@ -0,0 +1,138 @@
+    """
+    n4j_conn_id: str
+
+    template_fields = ['n4j_conn_id']
```
This doesn't do anything -- hooks aren't templated (just operators)

```suggestion
```

(plus remove the extra blank line)

ashb commented on airflow/providers/neo4j/hooks/neo4j_hook.py:
```diff
@@ -0,0 +1,138 @@
+        :rtype: dict
+        """
+        config: dict = {}
+        connection_object = Neo4JHook.get_connection(n4j_conn_id)
```
This should be a class method, not static, and then use `cls.get_connection`

Or perhaps better. Make this an instance method and look at `self.n4j_conn_id`

ashb commented on airflow/providers/neo4j/hooks/neo4j_hook.py:
```diff
@@ -0,0 +1,138 @@
+        )
+
+    @staticmethod
+    def get_session(driver: Driver) -> Session:
```
There is no point to this method -- please remove it.

ashb commented on airflow/providers/neo4j/hooks/neo4j_hook.py:
```diff
@@ -0,0 +1,138 @@
+        """
+        neo4j_config: dict = Neo4JHook.get_config(self.n4j_conn_id)
+        neo4j_driver: Driver = Neo4JHook.get_driver(neo4j_config)
+        neo4j_session: Session = Neo4JHook.get_session(neo4j_driver)
```
The more common pattern here would be something like:

```
neo4j_session = self.get_conn()
```

(or call it get_session)

And then that one method will do all of these three steps in one.

ashb commented on airflow/providers/neo4j/operators/neo4j_operator.py:
```diff
@@ -0,0 +1,90 @@
+# -*- coding: utf-8 -*-
```
Again, here I think this should be ` airflow/providers/neo4j/operators/neo4j.py` now.

ashb commented on docs/howto/operator/neo4j.rst:
```diff
@@ -0,0 +1,79 @@
+    under the License.
+
+Neo4j Operator
+==================
```
```suggestion
===========
```

ashb commented on docs/howto/operator/neo4j.rst:
```diff
@@ -0,0 +1,79 @@
+
+This operator enables Airflow DAGs to execute cypher queries against a Neo4j (or ONgDB) graph database.
+
+The Neo4J Operator extends :class:`airflow.models.baseoperator.BaseOperator`
```
```suggestion
```

Not needed - the auto rendered docs will tell us this (and all operators do)

ashb commented on docs/howto/operator/neo4j.rst:
```diff
@@ -0,0 +1,79 @@
+    :end-before: [END howto_operator_neo4j_from_file]
+
+Options
+^^^^^^^^^^^^^^^^^^
```
```suggestion
^^^^^^^
```

Header underlines should match the length of the line above.

ashb commented on setup.py:
```diff
@@ -279,6 +279,10 @@ def write_version(filename: str = os.path.join(*["airflow", "git_version"])):
 ]
+neo4j = [
+    'neo4j>=1.7.5',
+    'neobolt>=1.7.15'
```
What's this required for? Is it not a dependency of neo4j module?

ashb reviewed (changes requested):
Almost there!

Just a few file naming questions, and a relatively small change to the Hook to be more in line with the other hooks.

eladkal commented on airflow/providers/neo4j/example_dags/example_neo4j_operator.py:
```diff
@@ -0,0 +1,68 @@
+        cypher_query="input.cypher",
+        output_filename="output.csv",
+        soft_fail=True,
+        n4j_conn_id="ExampleN4J",
```
I wonder if this actually explain what is needed.
Only if you will look in the test code you will see that the conn type is `jdbc` and how to define it:
```
            Connection(
                conn_id='mock_config',
                conn_type='jdbc',
                host='localhost',
                port=1234,
                login='your_name_here',
                password='your_token_here'
            )
```

I think it should have its own connection available to define in the UI like other DB connections.

tfindlay-au replied:
I agree creating a custom connection type would be a good change. 

I used JDBC as it was the most generic way I could find to capture host/port/username/password but it is NOT actually a JDBC connection.

As I also noted, this does not leverage the DbApiHook or SqlAlchemy for the same reasons.

@eladkal I have not unpacked how to add a new connection type, so feel free to make your own contribtion here.

eladkal replied:
It should be added to:
https://github.com/apache/airflow/blob/master/airflow/utils/db.py#L62

tfindlay-tw committed:
Merge branch 'master' of https://github.com/apache/airflow

tfindlay-tw committed:
AIRFLOW-5920: Removed module file from example dags

tfindlay-tw committed:
AIRFLOW-5920: Removed unused template field definition.

tfindlay-au committed:
Update airflow/providers/neo4j/hooks/neo4j_hook.py

Co-Authored-By: Ash Berlin-Taylor <ash_github@firemirror.com>

tfindlay-tw committed:
AIRFLOW-5920: Removed get_session()
AIRFLOW-5920: Added get_conn()

tfindlay-au committed:
Apply suggestions from code review

Improve quality of documentation

Co-Authored-By: Ash Berlin-Taylor <ash_github@firemirror.com>

tfindlay-tw committed:
AIRFLOW-5920: Removed neobolt requirement as it is a dependency of neo4j

tfindlay-tw committed:
Merge remote-tracking branch 'origin/master'

tfindlay-tw committed:
Merge branch 'master' of https://github.com/apache/airflow

tfindlay-tw committed:
AIRFLOW-5920: Removed unused import, refactored test to match the change to the methods

tfindlay-tw committed:
AIRFLOW-5920: Removed example dag as advised, updated documentation.

tfindlay-tw committed:
AIRFLOW-5920: Renamed files and packages per AIP-21

tfindlay-tw committed:
Merge branch 'master' of https://github.com/apache/airflow

turbaszek commented on airflow/providers/neo4j/hook.py:
```diff
@@ -0,0 +1,133 @@
+        :return: dictionary with configuration values
+        :rtype: dict
+        """
+        config: dict = {}
```
Can you use `Dict` with proper types? It would add some additional information.

tfindlay-au replied:
Have changed to generic Dict, but had to type the value with Any as its mixed with a tuple and string. Let me know what you think.

turbaszek replied:
That's ok

turbaszek commented on airflow/providers/neo4j/hook.py:
```diff
@@ -0,0 +1,133 @@
+        connection_object = Neo4JHook.get_connection(n4j_conn_id)
+        if connection_object.login and connection_object.host:
+            config['credentials'] = connection_object.login, connection_object.password
+            config['host'] = "bolt://{0}:{1}".format(connection_object.host, connection_object.port)
```
Please use `os.path.join` for building URIs. Btw. we can use f-string in code ;)

tfindlay-au replied:
Used a f-string to join the host and port, but used `urllib.parse` to build a valid URL, `os.path.join` didnt really feel right for a URL. Let me know what you think, can simplify to just and f-string if you like.

turbaszek replied:
`urllib.parse` will be ok

turbaszek commented on airflow/providers/neo4j/hook.py:
```diff
@@ -0,0 +1,133 @@
+        config: dict = {}
+        connection_object = Neo4JHook.get_connection(n4j_conn_id)
+        if connection_object.login and connection_object.host:
+            config['credentials'] = connection_object.login, connection_object.password
```
```suggestion
            config['credentials'] = (connection_object.login, connection_object.password)
```
Just my opinion, I feel better with tuple brackets :)

turbaszek commented on airflow/providers/neo4j/operators/neo4j.py:
```diff
@@ -0,0 +1,90 @@
+        :param context:
+        :return: Row Count
+        :rtype: int
+        """
```
We rather do not use docstrings in `execute` method. All necessary information should be in class docstring.

tfindlay-au replied:
Done.

turbaszek reviewed (changes requested):
I really like the PR, is neat and clean 👌Just few nits and a question: would you mind to add example DAG that uses the new operator? It allows users to see how they can use it :)

tfindlay-tw committed:
Merge branch 'master' of https://github.com/apache/airflow

tfindlay-au committed:
Update airflow/providers/neo4j/hook.py

Looks good!

Co-Authored-By: Tomek Urbaszek <turbaszek@gmail.com>

tfindlay-tw committed:
Merge remote-tracking branch 'origin/master'

tfindlay-tw committed:
AIRFLOW-5920: Relocate documentation in neo4j.py
AIRFLOW-5920: Use urllib.parse to validate Bolt URL
AIRFLOW-5920: Move to dict -> typing.Dict

eladkal commented on airflow/providers/neo4j/operators/neo4j.py:
```diff
@@ -0,0 +1,86 @@
+from airflow.utils.decorators import apply_defaults
+
+
+class Neo4JOperator(BaseOperator):
```
I think the name is a bit confusing. The operator doesn't perform action on Neo4J (like running query, creating table etc..) but move data from Neo4J to CSV.  Usually such operators have naming convention of `x To y` Maybe `Neo4JOperatorDownloadOperator` ? (But that is just my thought maybe others will disagree)

tfindlay-au replied:
Sure, there were plans to make input & output operators and support different foramts. However I dont think its viable to develop a complete feature set in the first iteration.

I'll capture your thoughts in the JIRA ticket.

eladkal commented on airflow/providers/neo4j/operators/neo4j.py:
```diff
@@ -0,0 +1,86 @@
+        if result.peek() is None and self.soft_fail:
+            raise AirflowException("Query returned no rows")
+
+        return hook.to_csv(result, self.output_filename)
```
Currently this operator supports only CSV but I think it would be better if it would have param for `export_format` configurable by the user so it would be easier to add additional formats in the future.
Also like any CSV export you need to allow `field_delimiter`, `print_header` so users can configure how the csv output would look like.
Example:
https://github.com/apache/airflow/blob/v1-10-stable/airflow/contrib/operators/bigquery_to_gcs.py#L45-L50

tfindlay-au replied:
Yes, good feedback. Maybe in the future someone will implement this.

eladkal commented on airflow/providers/neo4j/operators/neo4j.py:
```diff
@@ -0,0 +1,86 @@
+
+    def execute(self, context):
+
+        if isfile(self.cypher_query):
```
Wouldn't it be easier to add `template_ext = ('.cql', '.cypher',)` to the operator?

eladkal commented on airflow/providers/neo4j/operators/neo4j.py:
```diff
@@ -0,0 +1,86 @@
+    n4j_conn_id = None
+    soft_fail = None
+
+    template_fields = ['cypher_query', 'output_filename', 'n4j_conn_id', 'soft_fail']
```
Why is `soft_fail` here? It's a boolean type what kind of templating you expect it to have?
